src/Sulu/Bundle/SecurityBundle/Controller/ResettingController.php

```diff
@@ -0,0 +1,309 @@
+    private function sendTokenEmail(User $user, $from, $to) {
+        if ($user->getPasswordResetToken() === null) {
+            throw new NoTokenFoundException($user);
+        }
+        if ($user->getTokenEmailsSent() === self::MAX_NUMBER_EMAILS) {
+            throw new TokenEmailsLimitReachedException(self::MAX_NUMBER_EMAILS, $user);
+        }
+        $mailer = $this->get('mailer');
+        $translator = $this->get('translator');
+        $em = $this->getDoctrine()->getManager();
+        $message = $mailer->createMessage()
+            ->setSubject(
+                $translator->trans(self::EMAIL_SUBJECT_KEY, array(), 'backend')
+            )
+            ->setFrom($from)
+            ->setTo($to)
+            ->setBody(
+                $translator->trans(self::EMAIL_MSG_KEY, array(), 'backend') . PHP_EOL .
+                $this->generateUrl('sulu_admin.reset', array('token' => $user->getPasswordResetToken()), true)
+            );
+        $mailer->send($message);
+        $user->setTokenEmailsSent($user->getTokenEmailsSent() + 1);
+        $em->persist($user);
+        $em->flush();
+    }
+
+    /**
+     * Changes the password of a user
+     * @param User $user
+     * @param string $password
+     * @throws MissingPasswordException
+     */
+    private function changePassword(User $user, $password) {
+        if ($password === '') {
+            throw new MissingPasswordException();
+        }
+        $em = $this->getDoctrine()->getManager();
+        $user->setPassword($this->encodePassword($user, $password, $user->getSalt()));
+        $em->persist($user);
+        $em->flush();
+    }
+
+    /**
+     * Generates a new token for a new user
+     * @param User $user
+     * @throws TokenAlreadyRequestedException
+     */
+    private function generateTokenForUser(User $user) {
+        if ($user->getPasswordResetToken() !== null &&
+            (new \DateTime())->add(self::getResetInterval()) < $user->getTokenExpiresAt()->add(self::getRequestInterval())
+        ) { // if a token was already requested within the request interval time frame
+            throw new TokenAlreadyRequestedException(self::getRequestInterval());
+        }
+        $em = $this->getDoctrine()->getManager();
+
+        $user->setPasswordResetToken($this->getToken());
+        $expireDateTime = (new \DateTime())->add(self::getResetInterval());
+        $user->setTokenExpiresAt($expireDateTime);
+        $user->setTokenEmailsSent(0);
+
+        $em->persist($user);
+        $em->flush();
+    }
+
+    /**
+     * Returns a unique token
+     * @return string - token
+     */
+    private function getToken() {
+        return $this->getUniqueToken($this->getTokenGenerator()->generateToken());
+    }
+
+    /**
+     * If the passed token is unique returns it back otherwise returns a unique token
+     * @param $startToken - the token to start width
+     * @return string a unique token
+     */
+    private function getUniqueToken($startToken) {
```

Comment:
The token-generator generates random token. It could happen (although the likelihood is quite small) that the token generated does already exists in the database. If this happens it would result in a conflict as a user needs to be identified by his token.
I solved this issue by creating a recursive function (getUniqueToken) which takes a token as input, looks if it's unique, if so returns it and otherwise calls itself recursively to get the unique token. The getToken-Method is just a wrapper for the sake of the readability of the code. i could also always write "$this->getUniqueToken($this->getTokenGenerator()->generateToken());" instead of "$this->getToken()"